src/broker/broker.rs

```diff
@@ -146,40 +140,23 @@ impl Broker {
 }
 
 impl Inner {
-    /// Continuously checks if the forwarding table needs to be rebuilt
+    /// Rebuild the forwarding table
     async fn rebuild_forwarding_table(&self) {
-        loop {
-            // sleep for 100ms, then check if the forwarding table is dirty
-            // and needs to be rebuilt
-            sleep(Duration::from_millis(100)).await;
-            if **self.forwarding_table_dirty.load() {
-                // ordering is very important to avoid race conditions here.
-                // the absolute first thing we need to do, is set dirty to
-                // false
-                //
-                // if we don't do this, then we can end up in a situation where
-                // while we're rebuilding the table, another thread sets it
-                // to be dirty again, then we set dirty to false, and never
-                // end up rebuilding the table.
-                self.forwarding_table_dirty.store(false.into());
-
-                // we need to make a clone of the clients values, so that we
-                // avoid any deadlock situations as we will be locking
-                // subscriptions on each client as well
-                let clients = self.clients
-                    .lock()
-                    .await
-                    .values()
-                    .cloned()
-                    .collect::<Vec<_>>();
-
-                // at this point it doesn't matter if another thread has set
-                // dirty = true in between the load and store lines above,
-                // because no matter what we're rebuilding the table
-                let forwarding_table = ForwardingTable::new(clients).await;
-                self.forwarding_table.store(forwarding_table.into());
-            }
-        }
+        let clients = self.clients
+            .lock()
+            .await
+            .values()
+            .cloned()
+            .collect::<Vec<_>>();
+
+        let new_table = Arc::new(
+            ForwardingTable::new(clients).await
+        );
+        
+        // use rcu to atomically update the forwarding table
+        self.forwarding_table.rcu(|_old_table| {
```

Comment:
This still contains a race condition, because we will become "eventually consistent" with a stale version of the rebuilt table.

The closure in the rcu method itself needs to lock clients and build the forwarding table.

Otherwise consider the scenario where at line 151, another thread calls rebuild_forwarding_table

Reply:
as discussed offline, since we cannot use an async closure here - use [block_in_place](https://docs.rs/tokio/latest/tokio/task/fn.block_in_place.html) in conjunction w/the blocking variants of the tokio mutex to reconstruct the forwarding table